stdlib/option.mc

```diff
@@ -1,4 +1,132 @@
 type Option
 
 con Some : Dyn -> Option
-con None : Option
\ No newline at end of file
+con None : Option
+
+-- Returns `true` if the option is a `Some` value.
+let isNone: Option -> Bool = lam o.
+  match o with None then
+    true
+  else
+    false
+
+-- Returns `true` if the option is a `None` value.
+let isSome: Option -> Bool = lam o.
+  match o with Some _ then
+    true
+  else
+    false
+
+-- Returns `true` if the option contains a value which
+-- satisfies the specified predicate.
+let contains: Option -> (Dyn -> Bool) -> Bool = lam o. lam p.
+  match o with Some t then
+    p t
+  else
+    false
+
+-- Applies a function to the contained value (if any).
+let map: Option -> (Dyn -> Dyn) -> Option = lam o. lam f.
+  match o with Some t then
+    Some(f t)
+  else
+    None
+    
+-- Applies a function to the contained value (if any),
+-- or returns the provided default (if not).
+let mapOr: Option -> Dyn -> (Dyn -> Dyn) -> Dyn = lam o. lam d. lam f.
+  match o with Some t then
+    f t
+  else
+    d
+
+-- Applies a function to the contained value (if any),
+-- or computes a default (if not).
+let mapOrElse: Option -> (Unit -> Dyn) -> Dyn = lam o. lam d.
+  match o with Some t then
+    t
+  else
+    d unit
+
+-- Returns `None` if either option is `None`, otherwise returns
+-- the first option.
+let and: Option -> Option -> Option = lam o1. lam o2.
+  match o1 with Some _ then
+    match o2 with Some _ then
+      o1
+    else
+      None
+  else
+    None
+
+-- Filters the contained value (if any) using the specified predicate.
+let filter: Option -> (Dyn -> Dyn) -> Option = lam o. lam p.
+  match o with Some t then
+    if p t then
+      o
+    else
+      None
+  else
+    None
+ 
+-- Returns the first option if it contains a value, otherwise returns
+-- the second option.
+let or: Option -> Option -> Option = lam o1. lam o2.
+  match o1 with Some _ then
+    o1
+  else
+    o2
+  
+-- Returns the option if it contains a value, otherwise calls the specified
+-- function and returns the result.
+let orElse: Option -> (Unit -> Option) -> Option = lam o. lam f.
+  match o with Some _ then
+    o
+  else
+    f unit
+
+-- Returns the first option if exactly one option is `Some`,
+-- otherwise returns `None`.
+let xor: Option -> Option -> Option = lam o1. lam o2.
+  match o1 with Some _ then
+    match o2 with None then
+      o1
+    else
+      None
+  else match o1 with None then
+    match o2 with Some _ then
+      o1
+    else
+      None
+  else
+    None
+
+-- Converts from `Option<Option<T>>` to `Option<T>`
+let flatten: Option -> Option = lam o.
+  match o with Some t then
+    t
+  else
+    None
+
+mexpr
+
+  utest isNone(None) with true in
+  utest isNone(Some(1)) with false in
+
+  utest isSome(Some(1)) with true in
+  utest isSome(None) with false in
+
+  utest contains(Some(1), 1) with true in
+  utest contains(Some(2), 1) with false in
+  utest contains(None, 1) with false in
```

Comment:
All of these function calls are missing the predicate `p`. Also, function calls are _curried_ in MCore, meaning that you pass them one by one rather than as a single tuple. What you want is probably something like this:
```
utest contains (Some 1) (lam n. eqi n 1) with true in
...
```